3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,222 @@
+import org.apache.maven.artifact.versioning.DefaultArtifactVersion
+
+// Scalaz
+import scalaz.Scalaz._
+import scalaz._
+
+/**
+ * PiiField trait. This corresponds to a configuration top-level field (i.e. either a POJO or a JSON field) along with
+ * a function to apply that strategy to a TransformMap
+ */
+sealed trait PiiField {
+  val strategy: PiiStrategy
+  val fieldName: String
+  def applyStrategy(fieldValue: String): String
+  val strategyApplication =
+    (v: Validation[String, String]) => v.flatMap(applyStrategy(_).success)
+  def transformer(transformMap: TransformMap) =
+    transformMap.collect {
+      case (inputField: String, (tf: Function2[String, String, Validation[String, String]], outputField: String))
+          if (outputField == fieldName) =>
+        (inputField, ((arg1: String, arg2: String) => tf.tupled.andThen(strategyApplication)((arg1, arg2)), outputField))
+    }
+}
+
+/**
+ * PiiStrategy trait. This corresponds to a strategy to apply to a single field. Currently only only String input is
+ * supported.
+ */
+trait PiiStrategy {
+  val scramble: String => String
+}
+
+/**
+ * Companion object. Lets us create a PiiPseudonymizerEnrichment
+ * from a JValue.
+ */
+object PiiPseudonymizerEnrichment extends ParseableEnrichment {
+  val supportedSchema = SchemaCriterion("com.snowplowanalytics.snowplow.enrichments", "pii_enrichment_config", "jsonscehma", 1, 0, 0)
+
+  def parse(config: JValue, schemaKey: SchemaKey): ValidatedNelMessage[PiiPseudonymizerEnrichment] = {
+    for {
+      conf      <- matchesSchema(config, schemaKey)
+      piiFields <- ScalazJson4sUtils.extract[List[JObject]](conf, "parameters", "pii").leftMap(_.getMessage)
+      strategyFunction <- ScalazJson4sUtils
+        .extract[String](config, "parameters", "strategy", "pseudonymize", "hashFunction")
+        .leftMap(_.getMessage)
+      hashFunction <- try {
+        MessageDigest.getInstance(strategyFunction).success
+      } catch {
+        case e: NoSuchAlgorithmException =>
+          s"Could not parse PII enrichment config: ${e.getMessage()}".fail
+      }
+      strategy = PiiStrategyPseudonymize(hashFunction)
+      piiFieldList <- piiFields.map {
+        case JObject(List(("pojo", JObject(List(("field", JString(fieldName))))))) => PiiPojo(strategy, fieldName).success
+        case JObject(
+            List(
+              ("json",
+               JObject(List(
+               ("field", JString(fieldName)),
+               ("schemaCriterion", JString(schemaCriterion)),
+               ("jsonPath", JString(jsonPath))))))) =>
+          SchemaCriterion
+            .parse(schemaCriterion)
+            .flatMap { sc => PiiJson(strategy, fieldName, sc, jsonPath).success }
+            .leftMap(_.getMessage)
+        case x => "Configuration file has unexpected structure".fail
+      }.sequenceU
+    } yield PiiPseudonymizerEnrichment(piiFieldList)
+  }.leftMap(getProcessingMessage(_)).toValidationNel
```

Comment:
`.leftMap(_.toProcessingMessageNel)`.

Reply:
I like that library. We should go ahead and replace it in all SCE